### PR TITLE
Feature/rails4

### PR DESCRIPTION
**Pull request by CGDS**
Opening early for review, still some cleanup for tests and the gemspec to stop using the rails project on my local.  @HonoreDB

### CODE REVIEW

File: lib/policy_machine_storage_adapters/active_record.rb [hunk 1]
```diff
@@ -292,22 +278,26 @@ def add_association(user_attribute, operation_set, object_attribute, policy_mach
     # If no associations are found then the empty array should be returned.
     #
     def associations_with(operation)
-      assocs = operation.policy_element_associations.all(include: [:user_attribute, :operations, :object_attribute])
-      assocs.map { |assoc| [assoc.user_attribute, Set.new(assoc.operations), assoc.object_attribute] }
+
+      assocs = operation.policy_element_associations(true).includes(:user_attribute, :operations, :object_attribute).all
+      assocs.map do |assoc|
+        assoc.clear_association_cache #TODO Either do this better (touch through HABTM on bulk insert?) or dont do this?
```
Comment (HonoreDB) [inline lib/policy_machine_storage_adapters/active_record.rb]: Does this undo the result of `includes` two lines up?

File: lib/policy_machine_storage_adapters/active_record.rb [hunk 2]
```diff
@@ -349,21 +339,26 @@ def is_privilege_single_policy_class(user_or_attribute, operation, object_or_att
...
-        associations_between(user_or_attribute, object_or_attribute).where(id: operation.policy_element_associations)
-      else
-        associations_between(user_or_attribute, object_or_attribute).joins(:operations).where(policy_elements: {unique_identifier: operation})
-      end
+      #Outstanding active record sql adapter prevents chaining an additional where using the association.
+      # TODO: fix when active record is fixed
       policy_classes_containing_object.all? do |pc|
-        base_scope.where(object_attribute_id: pc.ancestors).any?
+        if operation.is_a?(class_for_type('operation'))
+          associations_between(user_or_attribute, object_or_attribute).where(id: operation.policy_element_associations.to_a, object_attribute_id: pc.ancestors).any?
```
Comment (HonoreDB) [inline lib/policy_machine_storage_adapters/active_record.rb]: `operation.policy_element_associations.to_a` looks like a performance sadface, is it possible to keep this to a single SQL query?

File: lib/policy_machine_storage_adapters/active_record.rb [hunk 3]
```diff
@@ -349,21 +339,26 @@ def is_privilege_single_policy_class(user_or_attribute, operation, object_or_att
...
       end
     end
 
     # Pass in options to allow forced row ordering by id in results
     def scoped_privileges_single_policy_class(user_or_attribute, object_or_attribute, options = {})
       associations = associations_between(user_or_attribute, object_or_attribute).includes(:operations)
-      operations = associations.flat_map(&:operations).uniq
+      operations = associations.flat_map do |assoc|
+        assoc.clear_association_cache
+        assoc.operations
```
Comment (HonoreDB) [inline lib/policy_machine_storage_adapters/active_record.rb]: Does `assoc.operations(true)` work here?

File: spec/spec_helper.rb [hunk 4]
```diff
@@ -5,10 +5,12 @@
   add_filter 'test'
 end
 
-require_relative '../test/test_helper.rb'
+#require_relative '../test/test_helper.rb'
 
 require 'rspec'
 require 'pry'
+require 'policy_machine_test_app'
```
Comment (HonoreDB) [inline spec/spec_helper.rb]: This isn't checked in yet, but thank you for moya-izing this.

File: spec/support/shared_examples_shared_examples_policy_machine_storage_adapter.rb [hunk 5]
```diff
@@ -200,7 +200,7 @@
     it 'stores the association' do
       policy_machine_storage_adapter.add_association(@ua, Set.new([@r, @w]), @oa, 'some_policy_machine_uuid1')
       assocs_with_r = policy_machine_storage_adapter.associations_with(@r)
-      assocs_with_r.size == 1
+      assocs_with_r.size.should eq 1
```
Comment (HonoreDB) [inline spec/support/shared_examples_shared_examples_policy_machine_storage_adapter.rb]: *hangs head in shame*

File: lib/policy_machine_storage_adapters/active_record.rb (hunk 2 above)
Comment (CGDS) [inline lib/policy_machine_storage_adapters/active_record.rb]: I haven't found a way yet without fixing the AR postgres adapter, but Ill give it another look.  I could be wrong.  It is a sadface though.

File: lib/policy_machine_storage_adapters/active_record.rb (hunk 1 above)
Comment (CGDS) [inline lib/policy_machine_storage_adapters/active_record.rb]: yes and no.  Model cache yes, query cache no.  Maybe thats actually just a yes.  I'll verify.

File: spec/spec_helper.rb (hunk 4 above)
Comment (CGDS) [inline spec/spec_helper.rb]: It is a dead simple app.  I can push it to my repo and transfer it over when the paperwork clears, or I can ninja it into a branch of a private repo and then gitfu it when the paperwork clears.  Whacha think?

File: lib/policy_machine_storage_adapters/active_record.rb (hunk 3 above)
Comment (CGDS) [inline lib/policy_machine_storage_adapters/active_record.rb]: I dont know.  Good point.  Ill check.

File: spec/spec_helper.rb (hunk 4 above)
Comment (HonoreDB) [inline spec/spec_helper.rb]: I doubt anybody'd object if you just pushed it to your repo.

**Comment by CGDS** [comment]
@HonoreDB ready for your fork

**Comment by agodel** [comment]
running tests:
Finished in 30.17 seconds
665 examples, 0 failures, 12 pending
Coverage report generated for RSpec to /Users/agodel/Desktop/mdsol/pm/the_policy_machine/coverage. 677 / 758 LOC (89.31%) covered.
C02P44X7G3QD:the_policy_machine agodel$ bundle exec rspec test/
No examples found.

**Comment by HonoreDB** [comment]
Looks good to me, my only outstanding comments are performance-related so that can wait. Would you like me to merge?

**Comment by HonoreDB** [comment]
Or @agodel could merge since he doesn't have any commits on this branch.

File: spec/support/shared_examples_shared_examples_policy_machine_storage_adapter.rb (hunk 5 above)
Comment (agodel) [inline spec/support/shared_examples_shared_examples_policy_machine_storage_adapter.rb]: I thought should was deprecated? not block just surprised

File: test/tasks/setup.rake [hunk 6]
```diff
@@ -0,0 +1,20 @@
+namespace :pm do
+  namespace :test do
+    desc 'setup necessary scaffoliding for running policy machine specs'
+    task :prepare, [:db_type] do |t, args|
```
Comment (agodel) [inline test/tasks/setup.rake]: this is awesome

**Comment by agodel** [comment]
LGTM, @CGDS is this ready to merge or do you want others to have a crack at it?

File: spec/support/shared_examples_shared_examples_policy_machine_storage_adapter.rb (hunk 5 above)
Comment (HonoreDB) [inline spec/support/shared_examples_shared_examples_policy_machine_storage_adapter.rb]: It is, but it's all over the place in this gem, it'll be a separate story to convert all of it to expect.

**Comment by agodel** [comment]
merging

**Pull request merged**